### PR TITLE
Scraping from website

### PR DESCRIPTION
**Pull request by SuragNuthulapaty**
Added scraping of category, date, publication, blurb, all links in the blurb, and the article link for both bullet-pointed categories and non-bullet-pointed categories. Also added pandas and openpyxl for putting the data into a spreadsheet.

### CODE REVIEW

**Comment by derekeder** [comment]
@SuragNuthulapaty looking good! did you want a review of this pull request or should we review it together tonight?

**Comment by SuragNuthulapaty** [comment]
I was thinking we would review it later tonight

**Comment by herrerarobert** [comment]
The data in the spreadsheet seems good to me. One small comment: is there a built-in python lib we could use instead of importing pandas?

**Comment by SuragNuthulapaty** [comment]
@derekeder i just committed what we were working on yesterday if you want to take a look at it

File: scrape.py [hunk 1]
```diff
@@ -1,43 +1,219 @@
+from cgitb import strong
+from functools import cache
+from typing import final
```
Comment (derekeder) [inline scrape.py:3]: looks like these top 3 imports are unused. lets remove them

File: scrape.py [hunk 2]
```diff
@@ -1,43 +1,219 @@
...
 from bs4 import BeautifulSoup  #Importing the Beautiful Soup Library
 import requests				   #Importing the requests library
 import time					   #Importing the time library
```
Comment (derekeder) [inline scrape.py:6]: also unused

File: scrape.py [hunk 3]
```diff
@@ -1,43 +1,219 @@
...
-import csv					   #Importing the csv module
+import pandas as pd
+import json
+import requests_cache
 
+from zmq import curve_public
```
Comment (derekeder) [inline scrape.py:11]: also unused

File: scrape.py [hunk 4]
```diff
@@ -1,43 +1,219 @@
...
+
     for entry in entryContentEls:
-        print(entry.find_all('p'))
+
+        #each p tag is a category in each article
+        paragraphLinks = entry.find_all('p')
+
+        for p in paragraphLinks:
+            curItem = {}
+            # print(p.find('strong'))
```
Comment (derekeder) [inline scrape.py:80]: we should remove commented out print statements like these

File: scrape.py [hunk 5]
```diff
@@ -1,43 +1,219 @@
...
+                publication = 'Unknown'
+
+                #because of inconsistencies, the oublication can either be in an em or in an i tag, so check both
+                if p.find('em') is not None:
+                    publication = p.find('em').text.strip()[1:-1]
+                elif p.find('i') is not None:
+                    publication = p.find('i').text.strip()[1:-1]
+
+                curItem['publication'] = publication
+
```
Comment (derekeder) [inline scrape.py:133]: extra newlines here are unneeded

File: scrape.py [hunk 6]
```diff
@@ -1,43 +1,219 @@
...
+                links = []
+                for link in p.find_all('a'):
+                    links.append(link.get('href'))
+
+                curItem['links-within-blurb'] = links
+                curItem['article-link'] = digestLink
+
+                digestItems.append(curItem)
+
         #TODO should get Category, date, publication, blurb, link to the page, plus all links within the blurb, city, state
```
Comment (derekeder) [inline scrape.py:153]: this TODO is done, right? if so lets remove it

File: scrape.py [hunk 7]
```diff
@@ -1,43 +1,219 @@
...
+newArticles = []
+
+oldPostCounter = 0
+newPostCounter = 1
+
+#run until stop condition of no links
+while True:
+    response = requests.get(curentPosts.format(NUM_POSTS_PER_PAGE_NEW, newPostCounter))
+    parsedText = json.loads(response.text)
+    # soup = BeautifulSoup(parsedText, 'lxml')
```
Comment (derekeder) [inline scrape.py:174]: remove commented line

File: scrape.py [hunk 8]
```diff
@@ -1,43 +1,219 @@
...
+        link = metaEl.find_all('a', rel="bookmark")
+        for el in link:
+            digestLinks.append(el.get('href'))
 
+            
 # Get all the links for all of the pages, then extract info we are looking for
 for link in digestLinks:
     getDigestItems(link)
 
 # TODO store in a structured way.
```
Comment (derekeder) [inline scrape.py:216]: remove this TODO

**Review by derekeder — CHANGES_REQUESTED** [review]
@SuragNuthulapaty looking good! I made a few comments regarding commented lines and unused imports.

I also agree with @herrerarobert that we shouldn't need to import the entire `pandas` library to export to a spreadsheet. Its a pretty heavy library. For simplicity and also better compatibility, we should export as a `csv` file using the standard csv library.

We should write the csv using `csv.DictWriter` that will write our existing `digestItems` data to a csv file. Here's an example of how to do that: https://docs.python.org/3/library/csv.html#csv.DictWriter

Finally, we should include documentation in the Pull Request and also in the README on how to run the script and what the expected output should be. Also, any notes on how common parameters to change, how long it takes to run, etc. 

Here's an example of a pull request format you can follow: https://github.com/chihacknight/decarbonize-my-state/pull/85

**Review by derekeder — APPROVED** [review]
@SuragNuthulapaty this looks good! We can merge this in.

The next step is to review the output data and see if there's any changes to make to massage the data. I created a Google Sheet of the scraped digest items here as well as a summary table for counts of each category: https://docs.google.com/spreadsheets/d/1n_ruhDMz0g-b2yfhak1m9fq7Xr_0SY5rqPeamJqPHho/edit#gid=1843617235

From this, it looks like there are a few redundant / inconsistently named categories we're pulling like `OIL AND GAS` vs `OIL & GAS` etc: https://docs.google.com/spreadsheets/d/1n_ruhDMz0g-b2yfhak1m9fq7Xr_0SY5rqPeamJqPHho/edit#gid=934319479

We should try to clean up / merge these together. A few ideas for that:
- convert all text to uppercase and strip out all non-letter characters
- create a list of categories to ignore, like `YOUR AD HERE` and `YOUR MESSAGE HERE`
- for the `ALSO` and `MEANWHILE` articles, maybe associate them with the previously found category?

**Pull request merged**